viewing_party/party.py

```diff
@@ -1,23 +1,134 @@
-# ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    if title and genre and rating: 
+        return {"title": title, "genre": genre, "rating": rating}
+    else: 
+        return None
 
-# -----------------------------------------
-# ------------- WAVE 2 --------------------
-# -----------------------------------------
+def add_to_watched(user_data, movie):
+    
+    user_data["watched"].append(movie)
+    return user_data 
 
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
 
-# -----------------------------------------
-# ------------- WAVE 3 --------------------
-# -----------------------------------------
+def watch_movie(user_data, title):
+    for i in range(len(user_data["watchlist"])):
+        if user_data["watchlist"][i]["title"] == title:
+            user_data["watched"].append(user_data["watchlist"][i]["title"])
+            user_data["watchlist"].remove(user_data["watchlist"][i])
+    return user_data
 
+    
+def get_watched_avg_rating(user_data):
+    if len(user_data['watched']) == 0:
+        return 0
+    avg_total = 0
+    counter = 0
+    for dict in user_data["watched"]:
+        avg_total += dict['rating']
+        counter += 1
+    avg_rating = avg_total / counter
+    return avg_rating
+
+def get_most_watched_genre(user_data):
+    counter = {}
+    most_genre = ""
+    score = 0 
+    if len(user_data['watched']) == 0:
+        return None
+    for dict in user_data["watched"]:
+        if dict['genre'] not in counter:
+            counter[dict['genre']] = 1
+        else:
+            counter[dict['genre']] += 1
+    for key, values in counter.items():
+        if values >= score:
+            score = values
+            most_genre = key
+    return most_genre
+
+
+def get_unique_watched(user_data):
+    friends_list = []
+    user_list = []
+    diff_list = []
+    for i in range(len(user_data["watched"])):
+        user_list.append(user_data["watched"][i])
+    for j in range(len(user_data["friends"])):
+        for dict in (user_data["friends"][j]["watched"]):
+            friends_list.append(dict)
+    for dict in user_list: 
+        if dict not in friends_list:
+            diff_list.append(dict)
+    return diff_list

```

Comment:
`range` loops are useful we need to swap elements or rearrange items by id. In this case, we're solely focused on the values of the items so we can iterate through the list directly. This will clean up the syntax slightly by removing the i and j variables. 

```python
def get_unique_watched(user_data):
    friends_list = []
    user_list = []
    diff_list = []
    for movie in user_data["watched"]:
        user_list.append(movie)
    for friend in user_data["friends"]:
        for movie in friend["watched"]:
            friends_list.append(movie)
    for movie in user_list: 
        if movie not in friends_list:
            diff_list.append(movie)
    return diff_list

```

Reply:
Optionally, we could use list comprehension to build all of these lists! And we could move the logic to build the `friends_list` movie into a helper function since the next few functions need that logic. 

```python
def get_unique_watched(user_data):
    friends_list = []
    user_list = [movie for movie in user_data["watched"]]
    
    for friend in user_data["friends"]:
        for movie in friend["watched"]:
            friends_list.append(movie)

    return [ movie for movie in user_list if movie not in friends_list]

```